### PR TITLE
Separate universe minimization and evar normalization functions

### PR DESCRIPTION
**Pull request by SkySkimmer**
Depends on #6893.
Seems to break fiat-crypto through memory usage on travis: https://travis-ci.org/SkySkimmer/coq/jobs/349796288

### CODE REVIEW

**Comment by JasonGross** [comment]
Note that the file that fails in fiat-crypto, `RewriteAddToAdcInterp`, typically takes around 1 GB of RAM (e.g., it takes 932 328 ko in a [recent run of travis on master](https://travis-ci.org/mit-plv/fiat-crypto/jobs/350263654#L3956)), and in this PR, it dies with a peak usage of 5 922 896 ko.  I don't think that file stresses universes, but I could be wrong.  (Note, though, that it does deal with enormous a large number of branches resulting from an enormous term elaborated from a deeply nested `match` statement.)

**Comment by SkySkimmer** [comment]
OK I removed the problem commit and fiat passed on my travis.
Bench running at https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/397/console

**Comment by SkySkimmer** [comment]
Looks like I changed a few unsafe to constr into regular to_constr but needing abort_on_undefined:false.

**Comment by SkySkimmer** [comment]
```

┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                       │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   37.52   37.73 -0.56 % │   100896508682   101071659983 -0.17 % │   125818449599   125998352780 -0.14 % │  483380  483344 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  275.72  276.44 -0.26 % │   764122823873   766264138310 -0.28 % │  1081899062408  1084220482454 -0.21 % │ 1072776 1138372 -5.76 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  172.44  172.65 -0.12 % │   477540116358   478130395008 -0.12 % │   700087541469   700217942730 -0.02 % │  838872  838744 +0.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  185.52  185.64 -0.06 % │   512527246749   513266514013 -0.14 % │   689457642384   689635457039 -0.03 % │  643880  651868 -1.23 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 4014.73 4015.84 -0.03 % │ 11163795475409 11163792077858 +0.00 % │ 14580824977117 14576140613052 +0.03 % │ 2227232 2227544 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1439.73 1440.11 -0.03 % │  3999408198725  3998311701462 +0.03 % │  6657711520522  6659071994816 -0.02 % │ 1346724 1359228 -0.92 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   80.38   80.40 -0.02 % │   221786278358   221621791653 +0.07 % │   284993062346   285421581118 -0.15 % │  525056  526092 -0.20 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  201.88  201.89 -0.00 % │   558303636299   558447701693 -0.03 % │   765997557229   766241288937 -0.03 % │  847884  837964 +1.18 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-corn │ 1571.60 1571.52 +0.01 % │  4348498172709  4352016244075 -0.08 % │  6414020992705  6413969097416 +0.00 % │  847380  847316 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  460.20  460.12 +0.02 % │  1277623527768  1277663093090 -0.00 % │  2059293115839  2059165107941 +0.01 % │  814388  814300 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  704.08  703.58 +0.07 % │  1951492656823  1947299155397 +0.22 % │  2968084245171  2968134731920 -0.00 % │ 3375164 3375180 -0.00 % │   0   5 -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-math-classes │  229.06  228.86 +0.09 % │   629281180716   629528733153 -0.04 % │   787252012992   787401308284 -0.02 % │  531280  530992 +0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3111.86 3108.97 +0.09 % │  8632118520747  8624452286906 +0.09 % │ 13799898889931 13801905038038 -0.01 % │ 1241628 1241704 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   64.78   64.64 +0.22 % │   177893616912   178087989350 -0.11 % │   236122482815   236332236990 -0.09 % │  593868  593876 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  111.18  110.90 +0.25 % │   309765975813   308914693313 +0.28 % │   377127864375   376895305793 +0.06 % │  502192  502300 -0.02 % │   2   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   45.11   44.97 +0.31 % │   122391757067   122577990415 -0.15 % │   142715428996   142947942726 -0.16 % │  537740  537672 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   81.48   81.02 +0.57 % │   224845434029   224632446711 +0.09 % │   265505460230   265288565247 +0.08 % │  714608  713724 +0.12 % │   2   0  +nan % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘

```

**Comment by SkySkimmer** [comment]
Please review and merge.

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -1619,13 +1619,12 @@ let vernac_check_may_eval ~atts redexp glopt rc =
...
-  let c = EConstr.Unsafe.to_constr c in
   let sigma' = Evarconv.solve_unif_constraints_with_heuristics env sigma' in
   Evarconv.check_problems_are_solved env sigma';
-  let sigma',nf = Evarutil.nf_evars_and_universes sigma' in
+  let sigma' = Evd.minimize_universes sigma' in
   let uctx = Evd.universe_context_set sigma' in
   let env = Environ.push_context_set uctx (Evarutil.nf_env_evar sigma' env) in
-  let c = nf c in
+  let c = EConstr.to_constr ~abort_on_undefined_evars:false sigma' c in
   let j =
```
Comment (ppedrot) [inline vernac/vernacentries.ml]: You should probably seize the opportunity to move the broken `to_constr` call to the `else` branch below, it's an easy cleanup.

**Review by ppedrot — APPROVED** [review]
Except for the potential minor improvement suggested, I think this is fine and welcome. Any improvement on the crazy high-level universe API should be praised!

**Review by mattam82 — APPROVED** [review]
Nice cleanup, avoiding my kind of baroque `evar_map * (constr -> constr)` return type.

File: vernac/vernacentries.ml (hunk 1 above)
Comment (SkySkimmer) [reply vernac/vernacentries.ml]: Done.

File: engine/evarutil.mli [hunk 2]
```diff
@@ -178,11 +178,14 @@ val nf_evar_map_undefined : evar_map -> evar_map
...
 
 val nf_evars_and_universes : evar_map -> evar_map * (Constr.constr -> Constr.constr)
+[@@ocaml.deprecated "Use Evd.minimize_universes and nf_evars_universes"]
 val e_nf_evars_and_universes : evar_map ref -> (Constr.constr -> Constr.constr) * Universes.universe_opt_subst
+[@@ocaml.deprecated "Use Evd.minimize_universes and nf_evars_universes"]
 
 (** Normalize the evar map w.r.t. universes, after simplification of constraints.
     Return the substitution function for constrs as well. *)
 val nf_evar_map_universes : evar_map -> evar_map * (Constr.constr -> Constr.constr)
+[@@ocaml.deprecated "Use Evd.minimize_universes and nf_evar_map and nf_evars_universes"]
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply engine/evarutil.mli]: It's used in only a few places:
```
./vernac/obligations.ml:194:  let evm = Evarutil.nf_evar_map_undefined evm in
./vernac/classes.ml:189:  let sigma = Evarutil.nf_evar_map sigma in
./vernac/classes.ml:286:      let sigma = Evarutil.nf_evar_map sigma in
./vernac/classes.ml:290:      let sigma = Evarutil.nf_evar_map_undefined sigma in
./vernac/comProgramFixpoint.ml:196:  let sigma = Evarutil.nf_evar_map sigma in
./vernac/comProgramFixpoint.ml:261:  let evd = nf_evar_map_undefined evd in
```
nf_evar_info is used more (including by nf_enter).

**Comment by SkySkimmer** [comment]
>Note that the file that fails in fiat-crypto, RewriteAddToAdcInterp, typically takes around 1 GB of RAM (e.g., it takes 932 328 ko in a recent run of travis on master), and in this PR, it dies with a peak usage of 5 922 896 ko. I don't think that file stresses universes, but I could be wrong. (Note, though, that it does deal with enormous a large number of branches resulting from an enormous term elaborated from a deeply nested match statement.)

The change causing a failure AFAICT is making `Evarutil.nf_evars_universes` use `EConstr.to_constr` instead of `Universes.nf_evars_and_universes_opt_subst`.
Aside from universes they are different in that Universes normalises the evar arguments before substituting them in the body (ie Universes is call by value, EConstr is call by name).
Maybe we should try to have a common call by need implementation of evar substitution.

**Review by forestjulien — APPROVED** [review]

**Comment by mattam82** [comment]
Interesting @SkySkimmer! Maybe we should measure a cbv variant of to_constr as well. In most cases the substitutions are mainly variables so I don’t know what we would discover doing this...

**Comment by SkySkimmer** [comment]
Here's a go 
https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/403/console

Gaëtan Gilbert

**Comment by ppedrot** [comment]
@SkySkimmer So what do we do?

**Comment by SkySkimmer** [comment]
>@SkySkimmer So what do we do?

This is ready to merge modulo @amahboubi's review. I'll make another PR for the to_constr stuff.

**Comment by ppedrot** [comment]
I'm not sure it's a very good thing to merge a PR that introduces a performance regression without the corresponding fix...

**Comment by SkySkimmer** [comment]
This doesn't have a performance regression unless I forgot something though? Initially it did, breaking fiat-crypto through OOM, but I reverted that part of the PR.

**Comment by SkySkimmer** [comment]
fiat-crypto isn't compiling on any bench these days.
The important thing is that https://travis-ci.org/SkySkimmer/coq/jobs/367681419 succeeded.

**Comment by ppedrot** [comment]
The part of the code @amahboubi needs to review is rather small and simple. I don't think a proper review is needed, so let me merge this PR directly.

**Pull request merged**